### PR TITLE
Show uncacheable steps in UI

### PR DESCRIPTION
**Pull request by AkshitaB**
Fixes #

Changes proposed in this pull request:

- Uncacheable steps now visible in UI

Screen Shot 2022-02-16 at 4 16 11 PM

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/tango/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/tango/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: tango/local_workspace.py [hunk 1]
```diff
@@ -576,12 +570,7 @@ def registered_run(self, name: str) -> Run:
             raise KeyError(name)
         with SqliteDict(self.step_info_file, flag="r") as d:
             steps_for_run = {}
-            for step_symlink in run_dir.iterdir():
-                if not step_symlink.is_symlink():
-                    continue
-                step_name = str(step_symlink.name)
-                unique_id = str(step_symlink.resolve().name)
-                step_info = d[unique_id]
+            for step_name, step_info in d.items():
```
Comment (dirkgr) [inline tango/local_workspace.py:573]: This will run through all steps in sqlite. The old code only ran over all steps that are part of that run.

Comment (AkshitaB) [reply tango/local_workspace.py:573]: So, should we just create empty links for uncacheable steps?

File: tango/local_workspace.py [hunk 2]
```diff
@@ -574,18 +554,59 @@ def registered_run(self, name: str) -> Run:
...
+                except KeyError:
+                    d[step.unique_id] = StepInfo(
+                        step.unique_id,
+                        step.name if step.name != step.unique_id else None,
+                        step.__class__.__name__,
+                        step.VERSION,
+                        {dep.unique_id for dep in step.dependencies},
+                        step.cache_results,
+                    )
+                d.commit()
```
Comment (dirkgr) [inline tango/local_workspace.py:580]: Don't `commit()` every time. Just do it after all steps are done. I realize the old code did this, but fewer bigger transactions are going to be faster than lots of small ones.

File: tango/local_workspace.py [hunk 3]
```diff
@@ -574,18 +554,59 @@ def registered_run(self, name: str) -> Run:
...
+                unique_ids.append(step.unique_id)
+
+            run_step_info_file = self._run_step_info_file(name)
+            with open(run_step_info_file, "w") as file_ref:
+                json.dump(unique_ids, file_ref)
+
+    def _load_registered_run(self, name: str):
```
Comment (dirkgr) [inline tango/local_workspace.py:587]: Return type?

File: tango/local_workspace.py [hunk 4]
```diff
@@ -574,18 +554,59 @@ def registered_run(self, name: str) -> Run:
...
             raise KeyError(name)
-        with SqliteDict(self.step_info_file, flag="r") as d:
-            steps_for_run = {}
+        steps_for_run = self._load_registered_run(name)
+        return Run(name, steps_for_run, datetime.fromtimestamp(run_dir.stat().st_ctime))
+
+    def _run_step_info_file(self, name: str):
+        return self.runs_dir / name / "stepinfo.json"
+
+    def _save_registered_run(self, name: str, all_steps: Iterable[Step]):
```
Comment (dirkgr) [inline tango/local_workspace.py:563]: Return type?

**Review by dirkgr — CHANGES_REQUESTED** [review]
The same step can have a different name in different runs. For one experiment, a model might be the "final_model", and in another it might be the "baseline_model". It's bad design, my fault, that the `Step` and `StepInfo` classes even have a `step_name` member.

Because of this it is not safe to read the `step_name` out of the step info classes. If a different run ran in the meantime, a step's `step_name` could have changed. The step names have to stay with the run. It used to be the name of the symlink pointing to the result directory. Now, I recommend you make the json file containing the steps be a mapping, mapping step names to unique ids, to preserve this property.

**Pull request merged**